### ISSUE TITLE
Nested routes

### ISSUE THREAD

**Issue by porsager** [MithrilJS/mithril.js#2321]
### Description

Allowing nested routes as object to allow splitting route subsections in different files etc.

### Why

I copy the snippet below  between all my projects using `m.route`, and I've seen a lot of other people ask for it. (eg. on gitter)

### Possible Implementation & Open Questions

Support supplying a nested routes object directly to `m.route()`.

[flems example](https://flems.io/#0=N4IgzgpgNhDGAuEAmIBcICGAHLA6AVmCADQgBmAljEagNqgB2GAthGiLgBbzNQkiwA9g0Qj2IAL7FGLNugAeBIqSEiIY9AHcKDJIM255AAgC8RhhE1GACgCdB8gJ4AKYFKPAAOgyNGA5hDwqEbOWPZYxEbwGLYB8ACUpgB8IcBGAG4UlsHOiSYptN6+vszOAORM6WXERcW+2rr6uPYArohguMzYzrbJtXUl5RjVHkacthBkwbaRwrATGIjBzM2CbRC4UDoA1pFg8I4wwWVIFGBYUBiOqABGUIKw22XutvH9xW8+HzVfg2WcAEYRtFYoEjABqIxlIxYDABMqfXwAXSMEk+aP4kBgCAowhoIAAnKgAEySaQgJisdhIWw7DqEfiqUTwdgQeRYQS2eBGJCTDAtKDcrxfMoAemhAwGwUUGCgUAAIrSGNswABhQTMDkWEQ-Xxi1AUJASyXS3BgHR+GCKnbqzXCdTwXVQ0XzCCLCDG3ym13u63K21ah3eCSY6BweC4hj4gAsqABZJkVPQsK5WXpygEwmZrPZnO5vLI-MFHlqYs9kqMptlUGsMQOAftIjATv1hvLxWCwrqYvwgh07crRkUvZ0ta5jgb2sd7zFMAw6Q9JqHmzdC7H9Y1gZEtQkwdD2IjePYADY4wmKbJ2C1ILYlIysw6cxyuTy+QKhaXxRW6lW5QBVG81U3RtpxFUUDSNb9TXNBhLQgACIFsScHRbcDDVFMhaXUJAwDKU1MKyXQgLtKc91ILFw0jGMzykRM5EwHA7xUB8NBACg7Rfa9EKMTCNShXBRS42wym8djn25FMIwgMAePsZh+NFSS0xEhgxLzHklRVWS+LKASaTpFTvFUfYjFadpTBLMCBx-ZdNEWVChIHYIhNQpTpM9YI3ObT99OVXCpQ0ulg28bwVjMiBnD0WAWlYERcBuQQkEcSIy0iMhLngUQACU1naHpcuk+JPm8LBnAAeRufBw1wbYIEcMB8vWMAipK8rKuq2r6ucdLFmygqGvC5qWoYbwGj0AxBosiqqoQGq6oanrMvUHKmsa9phu8MgWgYHFhB4jK+tWwbIjCSYKGMMwygRSzfAmeAWlsHxpo6+a1sK5pkBaWAIucDBYFgSIungWBODyJJ3kG2ggZBpFcEySwjAAH0R0z+qhxYYdwYRoc4JHEYAAshnHYYmXREPeXwAH5RlwWm-oBoxaFOyhjEhYnpjR4nUQpwc0lp3B6cifnFsOvKiYxzgkROiYWYhIwccSXdfiMSI3HRbx701KhEPYG4MBuaB90oo90AAdhJY9z0pejcFgMAMyZR90EkKWQC2Py0HoC8kxAZgKHgcYqH4B6+HQbh4CwMBUFFQSGCwbY-FtjVRT9gPaSgAABAFcGz48U-9wOoCYkADiweiwHmCgsBZCQkQkIA)

```js
function flattenRoutes(routes, prefix = '') {
  return Object.keys(routes).reduce((acc, match) =>
    routes[match].view || routes[match].onmatch || routes[match].render
      ? { ...acc, [prefix + match]: routes[match] }
      : { ...acc, ...flattenRoutes(routes[match], prefix + match) }
    , {})
}
// usage
m.route(document.body, '/', flattenRoutes(routes))
```

### Is this something you're interested in working on?

Yes

**Comment by isiahmeadows** [comment]
@porsager If you added it to v2, you'd need to be careful to only consider properties starting with a `/` as a route. Also, this raises a few questions:

1. Should route objects and components intersect?
1. Should route objects and route resolvers intersect?
1. If the answer to 2 is "yes", what would be the expected behavior for `render`?
1. If the answer to 2 is "yes", what would be the expected behavior for `onmatch`?

The first three are probably pretty obvious: the first two should be allowed, and `render` should not be called on child match. The last one is much less obvious, because there's multiple useful ways you could go about it:

1. `onmatch` is called for all resolvers in the path and all awaited sequentially. The child route object could be optionally replaced or set on resolution, and if no child route matches, it falls back to using the parent itself as the end resolver.
1. `onmatch` is called for all resolvers in the path and all awaited in parallel, but only the child-most route would be rendered.
1. `onmatch` is called only for the child-most route to match.

The first would lend itself well to dynamic async route loading, but could lead to counter-intuitive behavior. The second would lend itself well to static async route loading, but would be simpler and slightly more consistent. The last is probably the simplest to implement and the simplest to understand: `"/child": child` &harr; `..._.mapKeys(child, c => "/child" + c)`.

**Comment by porsager** [comment]
I think you're misunderstanding my proposal.. I don't want any nested logic. Could you try to look again? I simply want to be able to give an object that has objects as values for keys, and if those objects aren't components or route resolves the keys are simply concatenated into the final route to match. (check the output from `flattenRoutes` in the flems).

**Comment by porsager** [comment]
Yeah, I'm not contemplating you should be able to mix paths with components or route resolvers. 

Merely a way to split out routes to allow structuring projects a little better. I don't even think it would be a breaking change.
